### ISSUE TITLE
Runtime SQLite query errors on the COLLATE NOCASE agents index bypass the canonical-fallback reconcile and still trigger spurious archive reconstruction (residual of the v0.3.13 integrity-guard fix)

### ISSUE THREAD

**Issue by bitjson** [Dicklesworthstone/mcp_agent_mail_rust#151]
Opus 4.8:

> **Summary**
> 
> v0.3.13 (commit `c5909c2a`) fixed the *periodic integrity-guard* false-positive: the bespoke
> ("frankensqlite") engine diverges from canonical SQLite on the `idx_agents_project_name_nocase`
> (`agents(project_id, name COLLATE NOCASE)`) unique index, and both the quick and full guard cycles
> now consult canonical SQLite via `reconcile_with_canonical(...)` before asserting corruption — which
> eliminated the chronic hourly false-reconstruct loop.
> 
> The same engine divergence still escalates to a spurious reconstruct on a path the `c5909c2a`
> reconcile does NOT cover. The periodic guard cycles and the runtime read-path health probe
> (`sqlite_primary_check_is_ok_with_canonical_fallback`) all reconcile against canonical SQLite, yet a
> live `ensure_project` (agent/project registration touching the `agents` table) still drove
> `live SQLite is suspect → degraded_read_only → reconstruct-from-archive` with no preceding
> canonical-reconcile log — while canonical `PRAGMA integrity_check` returns `ok` on the same file.
> Two candidate triggers (we did not isolate which):
> (1) a **write/mutate-path** corruption handler that escalates on the bespoke engine's spurious
> `database disk image is malformed: index idx_agents_project_name_nocase ...` error without the
> canonical second opinion; or
> (2) the reconcile is **fail-closed** and the separate canonical-probe connection cannot acquire the
> DB under the concurrent write + archive-git-commit load of registration, so it preserves the bespoke
> false-positive. Either way the result is an unnecessary reconstruct → `degraded_read_only` window →
> 30s `blocking dispatch timed out` on concurrent writes.
> 
> **Reproduction (observed)**
> 
> On a mailbox whose live `storage.sqlite3` the bespoke engine false-flags on the NOCASE index
> (canonical `sqlite3 'file:storage.sqlite3?mode=ro&immutable=1' 'PRAGMA integrity_check'` → `ok`),
> drive agent/project-registration traffic (e.g. register a new project + agent while other writers
> are active). Server log shows, with no preceding periodic-guard corruption WARN:
> 
> ```
> INFO live SQLite is suspect; read surfaces will prefer archive snapshots verdict_state=suspect durability_state=degraded_read_only
> INFO database reconstruction from archive complete ...
> ```
> 
> and concurrent `tools/call` writes return `-32603 ... Request timed out after 30s ... The database
> may be under heavy contention.` The periodic guard cycles reconcile and stay silent, so the
> escalation is a runtime-path event, not the periodic check.
> 
> **Proposed fix shape**
> 
> Gate every runtime corruption-verdict → recovery/`degraded_read_only` transition behind the same
> `reconcile_with_canonical(...)` second-opinion that `c5909c2a` added to the guard cycles (cover the
> write/mutate path, not only the read-path probe). If the cause is instead the fail-closed reconcile
> losing because the canonical probe cannot acquire the DB under concurrent write load, make
> "canonical probe could not run" a *retry/defer*, not an escalation to reconstruct. Ideal root fix:
> eliminate the bespoke NOCASE collation key-ordering divergence so the engine stops erroring on a
> structurally-valid index in the first place.
> 
> **Severity / scope**
> 
> Load-dependent, not rare: it fires whenever a live query hits the engine's NOCASE error, so on a
> busy mailbox under sustained agent/project registration it can recur frequently, each time producing
> a spurious `degraded_read_only` window + 30s write timeouts on concurrent operations. Worse, under
> concurrent load the reconstructs it triggers are a plausible origin for *real* index corruption (see
> the companion projects-index-corruption issue). Reproducible at `origin/main` @ `84f79b63` (v0.3.13);
> no fix has landed since `c5909c2a` (only the unrelated `f298fe5f`). Reproduction artifacts available
> on request.

**Comment by Dicklesworthstone** [comment]
I traced every runtime path that can flip the mailbox to `degraded_read_only` or fire an archive reconstruction, to find the NOCASE-corruption-class error that escapes the canonical-fallback reconcile. The short version: at `main` I can't find a path where a recoverable NOCASE query error still drives a reconstruct, because every reconstruct-triggering path is gated on a canonical second opinion *before* it does anything destructive. Details, so you can tell me if I'm missing the exact path you hit:

**1. The live write/tool path does not reconstruct in-process.** `ensure_project` → `queries::ensure_project` runs under `run_with_mvcc_retry`, which retries only MVCC/write-contention errors, never corruption. A corruption-class error there propagates up to `db_error_to_mcp_error` and surfaces as a terminal `DATABASE_CORRUPTION` MCP error (`retryable=false`); it never calls `recover_sqlite_file*`/`reconstruct_*`. The runtime corruption helper `DbPool::try_recover_from_corruption` has **no non-test caller**. So a live registration that hits the NOCASE error returns an error to the caller — it does not itself transition durability or reconstruct.

**2. Every actual reconstruct entry point already short-circuits on a canonical-healthy file.** `ensure_sqlite_file_healthy_with_archive` and `ensure_sqlite_file_healthy` both do `if sqlite_file_is_healthy(primary_path)? { return Ok(()) }` *before* touching the recovery admission controller. `sqlite_file_is_healthy` → `sqlite_primary_read_path_is_healthy` → `sqlite_primary_check_is_ok_with_canonical_fallback`, i.e. it consults canonical SQLite, so a NOCASE bespoke false-positive returns healthy here and the reconstruct is skipped. The startup probe (`run_startup_integrity_check`) and the periodic full cycle (`run_full_integrity_check`) both route through `reconcile_with_canonical` (the `c5909c2a` policy) on the same file. (This short-circuit and the canonical fallback are present at the SHA you reported, `84f79b63`.)

**3. The read-snapshot path can't surface a NOCASE error.** `open_read_db_pool` → `live_db_is_suspect` computes the verdict with `VerdictOptions::fast()`, which sets `skip_integrity_check` and `skip_db_sanity_check` true. The only DB-touching probes that then run are `db_exists` (stat) and `schema_populated` (which queries `sqlite_master` only — never the `agents` table or `idx_agents_project_name_nocase`). With no probe executing a query against that index, the verdict stays `Healthy` and `live_db_is_suspect` returns false. So this path won't reconstruct off a NOCASE error either.

Given the above, the `live SQLite is suspect ... degraded_read_only` + `reconstruction from archive complete` you observed during registration most plausibly came from the **periodic integrity guard or the startup probe firing concurrently** — both of which *are* canonical-reconciled — rather than the registration write path bypassing the reconcile. The companion #152 also showed your daemon was genuinely corrupting under load (canonical agreeing), so some of those reconstructs were real, not the NOCASE false-positive.

Two asks so I can land a targeted fix if there's still a gap:
- A server log slice showing the escalation **with the immediately-preceding lines** — specifically whether a `reconcile_with_canonical` / "canonical SQLite accepted/rejected" line (or its absence) precedes the `live SQLite is suspect` line. That tells us which of the two triggers you hypothesized actually fired.
- If it's trigger (2) — the fail-closed reconcile losing because the canonical probe couldn't acquire the DB under concurrent write + archive-git-commit load — that's a real, fixable defer-vs-escalate decision (a canonical probe that *errored* should defer/retry, not preserve the bespoke corruption verdict). The current `reconcile_with_canonical` preserves the corruption verdict when the canonical probe `Err`s, which is the conservative-but-wrong choice for a transient lock. If your logs show `canonical fallback could not run` lines, I'll change that arm to a bounded retry/defer.

Related: the loop-convergence half of #152 is now fixed on `main` (`e492b884`). Reproduction artifacts / the affected-DB copy you offered would help pin down whether #151 is still a live gap or fully covered by the existing canonical reconcile — happy to dig further with them.

**Comment by bitjson** [comment]
Opus 4.8:

> You asked whether a "canonical fallback could not run" line precedes the `live SQLite is
> suspect` escalation. On a busy multi-agent deployment it does, repeatedly — here's the slice.
> 
> Over one window the periodic full integrity guard against the **live** database logged:
> 
> ```
> WARN integrity probe rejected the file and canonical fallback could not run  path=.../storage.sqlite3   (31 occurrences)
> INFO live SQLite is suspect; read surfaces will prefer archive snapshots  durability_state=degraded_read_only
> INFO database reconstruction from archive complete  ... 0 parse errors
> ```
> 
> …cycling ~once a minute, while the same guard against the backup file logged the *overturned*
> case 22 times:
> 
> ```
> WARN primary sqlite integrity probe did not accept the file but canonical SQLite did; treating as healthy  path=.../storage.sqlite3.bak
> ```
> 
> So the bespoke probe is frequently rejecting files canonical SQLite would accept, and on the
> live DB the canonical second opinion routinely **can't acquire the busy database** (the
> daemon's own concurrent writers + the git-archive commit hold it), so the corruption verdict
> is preserved and drives a reconstruct.
> 
> This is the arm you flagged. At `main` @ `a893a56b`, `reconcile_with_canonical`
> (`crates/mcp-agent-mail-db/src/pool.rs:4783-4792`):
> 
> ```rust
> Err(canonical_error) => {
>     tracing::warn!(..., "integrity probe rejected the file and canonical fallback could not run");
>     Err(DbError::IntegrityCorruption { message, details })   // preserves the bespoke verdict
> }
> ```
> 
> Suggested fix (as you proposed): when the canonical probe `Err`s (could-not-run, i.e. a
> transient lock — distinct from canonical *rejecting* the file), bounded-retry/defer instead of
> preserving the corruption verdict, so a busy-DB lock isn't escalated to a reconstruct. The
> `Ok(false)` arm (canonical also rejects) should keep preserving it.
> 
> Note: a chunk of the bespoke false-positives feeding this likely originate engine-side — see
> the FrankenSQLite follow-up on frankensqlite#113, where `PRAGMA integrity_check` run against an
> actively-written DB falsely reports "page N is never used". Fixing this Err-arm defends against
> it regardless.

**Comment by Dicklesworthstone** [comment]
Confirmed the gap, and your log slice pins it exactly. This is the `Err(canonical_error)` arm of `reconcile_with_canonical` (`crates/mcp-agent-mail-db/src/pool.rs:4851-4861`):

```rust
Err(canonical_error) => {
    tracing::warn!(..., "integrity probe rejected the file and canonical fallback could not run");
    Err(DbError::IntegrityCorruption { message, details })   // preserves the bespoke verdict
}
```

On the live DB, the canonical second opinion (`sqlite_canonical_file_check_is_ok` → `CanonicalDbConn::open_file` + `PRAGMA …_check`) routinely **can't acquire the busy database** under concurrent writers + the git-archive commit, so it returns `Err`, and this arm preserves the bespoke false-positive corruption verdict → `degraded_read_only` → reconstruct. Your "31× `canonical fallback could not run`, cycling ~once/min, while the `.bak` probe is overturned 22×" slice is the smoking gun, and it lines up with the broader picture: this arm is *also* what makes the periodic full-cycle guard the most-plausible source of the registration-time reconstructs in my earlier trace (the live write path itself still doesn't reconstruct in-process — point 1 of that trace stands).

This is the arm I flagged I'd change. I'm holding the actual edit for a focused pass rather than landing it inline, because it's the corruption-recovery core and #152 shows real corruption can occur on this engine, so silencing a verdict here has to be done carefully. The design I intend:

1. **Distinguish "could not run" (transient lock) from "ran and failed".** The escalate-vs-defer decision must key on the `SqlError` kind: a busy/locked open (or a `PRAGMA` that bailed on a lock) is a transient-acquire failure and should **defer** (return the prior/healthy verdict, do not reconstruct); a canonical error that is *itself* a corruption/malformed signal should keep failing closed. Right now both collapse into the same `Err` arm.

2. **Bounded retry before deferring.** A short bounded retry/backoff on the canonical probe (it already has `open_sqlite_file_with_lock_retry` precedent on the primary side) so a momentary lock doesn't even reach the defer path.

3. **Consider an immutable read probe.** The strongest option is opening the canonical second-opinion connection `mode=ro&immutable=1` (the same way the repros run canonical `integrity_check`), which takes no locks at all and so isn't blocked by live writers — at the cost of possibly reading torn pages mid-write, which I need to weigh against the lock-defer approach. Likely: immutable-RO probe + treat a probe that still can't prove health as defer-not-escalate.

The `Ok(false)` arm (canonical *also* rejects — i.e. real corruption, as in #152) stays fail-closed and keeps preserving the verdict; only the could-not-run arm changes.

Note (as you flagged): some of the bespoke false-positives feeding this originate engine-side — `frankensqlite#113` (`PRAGMA integrity_check` against an actively-written DB falsely reporting "page N is never used"). Fixing this Err-arm defends against that regardless, so it's the right local fix even with the engine work pending.

Leaving this open as the tracking issue for that defer-vs-escalate change; the loop-convergence half of the related #152 is already fixed (`e492b884`).

**Comment by Dicklesworthstone** [comment]
Fixed in 89d3d8fe.

This closes the runtime-path gap left after v0.3.13 (c5909c2a covered the periodic guard + read-path probe, but not the case where the canonical second-opinion probe could not *run*).

`reconcile_with_canonical` (pool.rs) previously preserved the bespoke `IntegrityCorruption` verdict whenever the canonical probe returned an `Err`, conflating two distinct cases:
1. canonical positively rejected the file (`Ok(false)`) → stay fail-closed (correct, unchanged); and
2. the canonical probe could not acquire the DB under sustained concurrent write + archive-git-commit load → previously also preserved the verdict, which let a divergent-engine false positive (the `idx_agents_project_name_nocase` COLLATE NOCASE "malformed" report that canonical otherwise calls `ok`) escalate straight to an archive reconstruct → `degraded_read_only` window that timed out every concurrent write.

Now a *lock/busy* (`is_lock_error`), *stale-wal/snapshot-conflict*, or *transient-recovery* canonical-probe failure is demoted to a neutral deferral error — asserted **non-corruption AND non-recovery** by a regression test, so it cannot re-trip `is_corruption_error_message` / `is_sqlite_recovery_error_message` in the integrity guard and re-escalate the very reconstruct it is deferring. The guard re-probes on the next cycle once contention clears. A genuine canonical rejection (`Ok(false)`) and a *non-transient* probe failure (e.g. generic open/config error) still preserve the corruption verdict (existing fail-closed test retained and clarified).

Test: `reconcile_canonical_defers_when_probe_blocked_by_lock_contention` (covers `database is locked`, `database table is locked`, snapshot-conflict; asserts the deferral message is self-describing, non-corruption, and non-recovery).

The ideal root fix — eliminating the bespoke NOCASE collation key-ordering divergence in the engine so it stops false-flagging a structurally-valid index — remains a FrankenSQLite-side concern; this change ensures that false flag can no longer trigger a spurious recovery on this side.
